File: packages/sdk-ts/src/resources/tools.ts
/**
 * Tools resource — list toolkits, inspect tools.
 */

import type { HttpClient } from '../http.js';
import type { PageResult, Toolkit, Tool } from '../types.js';

export interface ListToolkitsOptions {
  q?: string;
  cursor?: string;
  limit?: number;
}

export interface ListToolsOptions {
  cursor?: string;
  limit?: number;
}

export class ToolsResource {
  constructor(private readonly http: HttpClient) {}

  /**
   * List available toolkits, optionally filtered by a full-text search query.
   */
  listToolkits(opts: ListToolkitsOptions = {}): Promise<PageResult<Toolkit>> {
    return this.http.get('/toolkits', {
      q: opts.q,
      cursor: opts.cursor,
      limit: opts.limit,
    });
  }

  /**
   * List tools for a specific toolkit.
   */
  listTools(toolkitSlug: string, opts: ListToolsOptions = {}): Promise<PageResult<Tool>> {
    return this.http.get(`/toolkits/${toolkitSlug}/tools`, {
      cursor: opts.cursor,
      limit: opts.limit,
    });
  }

  /**
   * Get a specific tool by toolkit slug and tool slug.
   */
  getTool(toolkitSlug: string, toolSlug: string): Promise<Tool> {
    return this.http.get(`/toolkits/${toolkitSlug}/tools/${toolSlug}`);
  }
}

File: packages/sdk-py/arcane/resources/tools.py
"""Tools and toolkits resource."""

from __future__ import annotations

from typing import Any

from ..http import HttpClient, AsyncHttpClient
from ..types import PageResult, Toolkit, Tool


class ToolsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list_toolkits(
        self,
        *,
        q: str | None = None,
        cursor: str | None = None,
        limit: int | None = None,
    ) -> PageResult[Toolkit]:
        """List toolkits, optionally filtered by a search query."""
        data = self._http.get(
            "/toolkits",
            params={"q": q, "cursor": cursor, "limit": limit},
        )
        return PageResult.from_dict(data, Toolkit)

    def list_tools(
        self,
        toolkit_slug: str,
        *,
        cursor: str | None = None,
        limit: int | None = None,
    ) -> PageResult[Tool]:
        """List tools for a given toolkit."""
        data = self._http.get(
            f"/toolkits/{toolkit_slug}/tools",
            params={"cursor": cursor, "limit": limit},
        )
        return PageResult.from_dict(data, Tool)

    def get_tool(self, toolkit_slug: str, tool_slug: str) -> Tool:
        """Get a specific tool by toolkit and tool slug."""
        data = self._http.get(f"/toolkits/{toolkit_slug}/tools/{tool_slug}")
        return Tool.from_dict(data)


class AsyncToolsResource:
    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def list_toolkits(
        self,
        *,
        q: str | None = None,
        cursor: str | None = None,
        limit: int | None = None,
    ) -> PageResult[Toolkit]:
        data = await self._http.get(
            "/toolkits",
            params={"q": q, "cursor": cursor, "limit": limit},
        )
        return PageResult.from_dict(data, Toolkit)

    async def list_tools(
        self,
        toolkit_slug: str,
        *,
        cursor: str | None = None,
        limit: int | None = None,
    ) -> PageResult[Tool]:
        data = await self._http.get(
            f"/toolkits/{toolkit_slug}/tools",
            params={"cursor": cursor, "limit": limit},
        )
        return PageResult.from_dict(data, Tool)

    async def get_tool(self, toolkit_slug: str, tool_slug: str) -> Tool:
        data = await self._http.get(f"/toolkits/{toolkit_slug}/tools/{tool_slug}")
        return Tool.from_dict(data)
